### ISSUE TITLE
Merge sys-mor into master

### ISSUE THREAD

**Issue by sdrave** [pymor/pymor#412]
Before merging sys-mor into master, the following has to be done:

- [x] Make some version of pymess publicly available. (Could also be an inofficial version packaged with pyMOR. Using it could raise some warning about experimental code.)
- [x] Make pymess available on travis.
- [x] Make all tests pass on travis.
~~[ ] Add a basic sys-mor example which uses pyMOR's discretization toolkit.~~

**Comment by sdrave** [comment]
I will see to the example when everything else is ready ..

**Comment by sdrave** [comment]
pymess is now publicly available! There are supposed to be pip-installable packages soon. Then we can proceed with integrating pymess in our travis env.

**Comment by pmli** [comment]
The wheels are here: http://www.mpi-magdeburg.mpg.de/projects/mess

**Comment by sdrave** [comment]
@renemilk, could you add the pymess wheel to our travis setup? (The wheels are also available via https, BTW.)

**Comment by renemilk** [comment]
#451 Yes, I could.

**Comment by sdrave** [comment]
@pmli, could you cherry-pick 08bfb233ba4dfbe32d5cea7a78706c257b2d1417 from #451 into sys-mor? This should add pymess on travis.

**Comment by pmli** [comment]
> @pmli, could you cherry-pick 08bfb23 from #451 into sys-mor? This should add pymess on travis.

Done.
Also, I rebased the branch in #433.

**Comment by sdrave** [comment]
@pmli, could you also cherry-pick 90745d8665600ba67e5f02173bfbc691d0bed6b1? There was an error in the url for the pymess wheels ...

**Comment by sdrave** [comment]
@renemilk, now pymess installs, but pymor does not seems to find it: https://travis-ci.org/pymor/pymor/builds/405352219. Cannot reproduce locally, could you check in the docker container?

@pmli, slycot fails to install since we require `slycot==0.3.1` which is unavailable. `slycot==0.3.3` is, however. I guess it's ok to require `slycot>=0.3.3`?

**Comment by pmli** [comment]
@sdrave, the thing is that `slycot.sg03ad` has a bug with checking the size of `ldwork` in versions `0.3.2` and `0.3.3`. In 20ce300962fc8f44fe4722564f4bed9e86830ba2, I changed the `job` argument from `X` to `B`, which works for all versions. This change makes `slycot.sg03ad` take more time and memory, because it additionally estimates the condition number of the Lyapunov operator, but I guess this could also be useful. I'll report this bug when I find more free time, and then we could see how to use `slycot.sg03ad` in future versions.

**Comment by sdrave** [comment]
Great, everything seems to work now. Just pushed a commit to `sys-mor` which ensures tests to fail if slycot or pymess are missing on travis instead of silently skipping the tests.

**Comment by sdrave** [comment]
Regarding tests, we do not seem to have any system tests yet for the `sys-mor` branch. I guess we should add something to `pymortests.demos`. (I know that we are planning to do everything in notebooks and test those, but the testing infra will not happen before next year ..)

**Comment by pmli** [comment]
There are `heat.py` and `delay.py` in `pymordemos`, although they are not as fancy as other demos.

**Comment by sdrave** [comment]
I have added both to `pymortests.demos`. I will be on vacation now for two weeks. After that I would propose to think of really merging this into master. It would be really good if we had a pyMOR 0.5 release this year, and I would like to see `sys-mor` in master for some time before the release. @pmli, is there anything left to do from your side before merging this?

**Comment by sdrave** [comment]
@pmli, I have weired issue regarding pymess wheels and slycot:

```python
import slycot
import pymess
```

works fine. However

```
import pymess
import slycot
```

leads to

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/stephan/.virtualenvs/pymor/lib/python3.5/site-packages/slycot/__init__.py", line 16, in <module>
    from .analysis import ab01nd,ab05md,ab05nd,ab07nd,ab08nd, \
  File "/home/stephan/.virtualenvs/pymor/lib/python3.5/site-packages/slycot/analysis.py", line 21, in <module>
    from . import _wrapper
ImportError: /home/stephan/.virtualenvs/pymor/lib/python3.5/site-packages/pymess/.libs/libgfortran.so.3: version `GFORTRAN_1.4' not found (required by /home/stephan/.virtualenvs/pymor/lib/python3.5/site-packages/slycot/_wrapper.cpython-35m-x86_64-linux-gnu.so)
```

is this a known issue? One workaround would be to first try to import `slycot` in the `HAVE_PYMESS` check in `pymor.core.config`.

**Comment by sdrave** [comment]
I opened a separate issue for the PDE example. This can also happen after merging.

**Comment by pmli** [comment]
@sdrave Ok, thanks. I'll be away for three weeks.
I think we had discussed that parameterizing `LTISystem` class would be a good idea. Also, merging #433 would be nice.

I cannot reproduce the issue you found, with pymess wheels or pymess compiled from source. Both orders of importing work for me. No idea what the message about `GFORTRAN_1.4` could mean. Do you have `gfortran` installed on your system?

**Comment by sdrave** [comment]
@drittelhacker, nope this is on stretch in a virtualenv. I can confirm that the `LD_PRELOAD` hack works here as well.

The `pymess` wheel ships its own libgfortran whereas my slycot package uses the system libgfortran. I always thought that python DL_OPENs C extension modules in a way that dependent shared libraries are private to the respective extension. Obviously I am mistaken. And somehow there seems to be an issue with the lib in the pymess wheel.

**Comment by sdrave** [comment]
@drittelhacker, so I have learned a few new things:

Python DLOPENS with `RTLD_LOCAL | RTLD_NOW`. So symbols from the libgfrotran from pymess should not influence later loaded libs. *However*, the runtime loader never loads two different libraries with the same SONAME. So the first loaded libgfortran is always used for both `slycot` and `pymess`. Now, the libgfortran shipped with pymess seems to be older, only having GFORTRAN_1.1 as version symbol, whereas my system libgfortran has up to GFORTRAN_1.6.

So the issue is that this libgfortran in the pymess wheel has a standard SONAME. However, libs shipped in wheels should have a unique non-standard SONAME, e.g.

```/home/stephan/.virtualenvs/pymor/lib/python3.5/site-packages/numpy/.libs/libgfortran-ed201abd.so.3.0.0```

I believe, the standard toolchain for creating manylinux wheels takes care of this. I believe, last time I was in Magdeburg, someone said that they were unable to use this, however ...

**Comment by drittelhacker** [comment]
@sdrave Say thank you to the wheels rules. Actually you are not supposed to ship libgfortran. This is probably due to the reasons you figured out now. However, gfortran is not part of a default install in most linux distros, so not shipping any libgfortran is not a real option. As a quick fix we are working on an exchange of the so-name to avoid conflicts in the 1.0.1 release of C-M.E.S.S/Py-M.E.S.S..

The problem with the version symbols is related to the stone-age gcc-suites in the CentOS installations requested for the manylinux builds. For the near future, we hope for PEP 571   https://www.python.org/dev/peps/pep-0571/ to give us more flexibility in shipping proper compliant wheels.

I will drop you a line once the quick-fix wheels are available.

**Comment by sdrave** [comment]
@pmli:
- Regarding the parameterization of `LTISystem` it would be a good idea to open a separate issue for that. One questions would be where the parameterization should be considered. We can still do this when the code is in master.

- I guess it makes sense to merge #433 before merging this into master. Do that whenever you like.

Once you are ready to merge with master, I will take another look at sys-mor ..

**Comment by pmli** [comment]
@sdrave I've merged #433. I think the issue of linear and antilinear operators should be resolved, in particular the discussion in #448, before merging sys-mor into master.

**Comment by sdrave** [comment]
Done in #464.
